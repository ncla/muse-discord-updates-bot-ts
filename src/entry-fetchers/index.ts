import {Update} from "@/src/updates";

export interface EntryFetcher {
    fetch(): Promise<Update[]>;
}