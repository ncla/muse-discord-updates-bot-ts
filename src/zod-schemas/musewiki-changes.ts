import { z } from "zod";

// Zod schema for MediaWiki recent change
export const MediaWikiRecentChangeSchema = z.object({
    type: z.string(),
    ns: z.number(),
    title: z.string(),
    pageid: z.number(),
    revid: z.number(),
    old_revid: z.number(),
    rcid: z.number(),
    user: z.string(),
    oldlen: z.number().optional(),
    newlen: z.number().optional(),
    comment: z.string().optional(),
    parsedcomment: z.string().optional(),
    new: z.string().optional(),
    minor: z.string().optional(),
    logtype: z.string().optional(),
    logaction: z.string().optional(),
    timestamp: z.string()
});

// Zod schema for MediaWiki recent changes response
export const MediaWikiRecentChangesResponseSchema = z.object({
    batchcomplete: z.string(),
    continue: z.object({
        rccontinue: z.string(),
        continue: z.string()
    }).optional(),
    query: z.object({
        recentchanges: z.array(MediaWikiRecentChangeSchema)
    })
});