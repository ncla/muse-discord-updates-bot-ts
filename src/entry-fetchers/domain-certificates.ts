import {EntryFetcher} from "@/src/entry-fetchers/index";
import {z} from "zod";
import {createBlankUpdate, DomainCertificateUpdate, UpdateType} from "@/src/updates";

const CertificateSchema = z.object({
    issuer_ca_id: z.number(),
    issuer_name: z.string(),
    common_name: z.string(),
    name_value: z.string(),
    id: z.number(),
    entry_timestamp: z.string(),
    not_before: z.string(),
    not_after: z.string(),
    serial_number: z.string(),
    result_count: z.number(),
})

type Certificate = z.infer<typeof CertificateSchema>

export class DomainCertificates implements EntryFetcher
{
    private domain: string | undefined;

    constructor(domain: string | undefined) {
        this.domain = domain
    }

    async fetch()
    {
        if (this.domain === undefined) {
            throw new Error('Domain is not set')
        }

        const url = new URL(`https://crt.sh/json`);
        url.searchParams.append('q', this.domain)

        const response = await fetch(url.toString());

        if (!response.ok) {
            throw new Error(`Response status: ${response.status}`);
        }

        const json = await response.json();

        const CertificateArraySchema = z.array(CertificateSchema)

        const parsedJson = CertificateArraySchema.parse(json)

        const uniqueCommonNames = new Set(
            parsedJson
                .map((cert: Certificate) => cert.name_value.split('\n'))
                .flat()
        )

        return Array.from(uniqueCommonNames).map(commonName => this.mapCertificateToUpdateEntry(commonName))
    }

    private mapCertificateToUpdateEntry(domain: string): DomainCertificateUpdate
    {
        return {
            ...createBlankUpdate(),
            type: UpdateType.DOMAIN_CERTIFICATE,
            id: domain,
            uniqueId: domain,
        }
    }
}