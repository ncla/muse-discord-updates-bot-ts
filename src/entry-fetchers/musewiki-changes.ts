import { EntryFetcher } from "@/src/entry-fetchers/index";
import { MuseWikiChangeUpdate, UpdateType } from "@/src/updates";
import {MediaWikiRecentChangesResponseSchema} from "@/src/zod-schemas/musewiki-changes";

export class MuseWikiChanges implements EntryFetcher {
    private readonly apiUrl: string = 'https://musewiki.org/api.php';
    private readonly apiParams = {
        action: 'query',
        list: 'recentchanges',
        rclimit: '50',
        rcprop: 'title|ids|sizes|flags|user|comment|parsedcomment|loginfo|timestamp',
        format: 'json'
    };

    async fetch(): Promise<MuseWikiChangeUpdate[]> {
        const url = new URL(this.apiUrl);

        Object.entries(this.apiParams).forEach(([key, value]) => {
            url.searchParams.append(key, value);
        });

        const response = await fetch(url.toString());

        if (!response.ok) {
            throw new Error(`HTTP error! Status: ${response.status}`);
        }

        const data = await response.json();

        const validatedData = MediaWikiRecentChangesResponseSchema.parse(data);
        const recentChanges = validatedData.query.recentchanges;

        console.log(`Fetched ${recentChanges.length} recent changes from MuseWiki`);

        return recentChanges.map((change): MuseWikiChangeUpdate => {
            const changeType = change.type === 'new' ? 'new' : (change.logtype ? 'log' : 'edit');

            return {
                type: UpdateType.MUSEWIKI_CHANGE,
                uniqueId: `${change.pageid}-${change.revid}`,
                id: `${change.revid}`,
                title: change.title,
                change_type: changeType,
                user: change.user,
                comment: change.comment || null,
                pageid: change.pageid,
                old_revid: change.old_revid,
                oldlen: change.oldlen,
                newlen: change.newlen,
                created_at: new Date(change.timestamp),
            };
        });
    }
}
