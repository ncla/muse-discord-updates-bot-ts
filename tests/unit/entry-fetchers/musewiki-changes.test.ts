import { expect, test } from 'vitest'
import { MuseWikiChanges } from '@/src/entry-fetchers/musewiki-changes'
import { http, HttpResponse } from 'msw'
import { setupServer } from 'msw/node'
import { promises as fs } from 'fs'
import path from 'node:path'

test('throws error on not OK response', async () => {
    const requestHandlers = [
        http.get('https://musewiki.org/api.php', () => {
            return HttpResponse.json(
                {},
                {
                    status: 503,
                }
            )
        }),
    ]

    const server = setupServer(...requestHandlers)
    server.listen({ onUnhandledRequest: 'error' })

    const fetcher = new MuseWikiChanges()
    await expect(fetcher.fetch()).rejects.toThrow('HTTP error! Status: 503')

    server.close()
    server.resetHandlers()
})

test('throws error on failed zod validation', async () => {
    const invalidResponse = {
        invalid: 'data'
    }

    const requestHandlers = [
        http.get('https://musewiki.org/api.php', () => {
            return HttpResponse.json(
                invalidResponse,
                {
                    status: 200,
                }
            )
        }),
    ]

    const server = setupServer(...requestHandlers)
    server.listen({ onUnhandledRequest: 'error' })

    const fetcher = new MuseWikiChanges()
    await expect(fetcher.fetch()).rejects.toThrow()

    server.close()
    server.resetHandlers()
})

test('returns updates on successful response', async () => {
    const goodJsonResponse = await fs.readFile(
        path.join(__dirname, '../../__fixtures__/entry-fetchers/musewiki-changes/good-response.json'),
        { encoding: 'utf-8' }
    )
    const validResponse = JSON.parse(goodJsonResponse)

    const requestHandlers = [
        http.get('https://musewiki.org/api.php', () => {
            return HttpResponse.json(
                validResponse,
                {
                    status: 200,
                }
            )
        }),
    ]

    const server = setupServer(...requestHandlers)
    server.listen({ onUnhandledRequest: 'error' })

    const fetcher = new MuseWikiChanges()
    const result = await fetcher.fetch()

    expect(result).toMatchSnapshot()

    server.close()
    server.resetHandlers()
})