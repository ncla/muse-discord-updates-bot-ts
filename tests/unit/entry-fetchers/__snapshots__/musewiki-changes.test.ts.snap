// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`returns updates on successful response 1`] = `
[
  {
    "change_type": "edit",
    "comment": null,
    "created_at": 2025-06-07T14:20:07.000Z,
    "id": "219421",
    "newlen": 1194,
    "old_revid": 219419,
    "oldlen": 1144,
    "pageid": 35995,
    "title": "Chromes",
    "type": "MUSEWIKI_CHANGE",
    "uniqueId": "35995-219421",
    "user": "HyperChondriacMuser",
  },
  {
    "change_type": "log",
    "comment": null,
    "created_at": 2025-06-07T14:19:06.000Z,
    "id": "219420",
    "newlen": 0,
    "old_revid": 0,
    "oldlen": 0,
    "pageid": 36081,
    "title": "File:Chromes the one.jpg",
    "type": "MUSEWIKI_CHANGE",
    "uniqueId": "36081-219420",
    "user": "HyperChondriacMuser",
  },
  {
    "change_type": "edit",
    "comment": null,
    "created_at": 2025-06-07T14:16:45.000Z,
    "id": "219419",
    "newlen": 1144,
    "old_revid": 219417,
    "oldlen": 1117,
    "pageid": 35995,
    "title": "Chromes",
    "type": "MUSEWIKI_CHANGE",
    "uniqueId": "35995-219419",
    "user": "HyperChondriacMuser",
  },
  {
    "change_type": "edit",
    "comment": null,
    "created_at": 2025-06-07T14:16:04.000Z,
    "id": "219418",
    "newlen": 1910,
    "old_revid": 219288,
    "oldlen": 1876,
    "pageid": 35999,
    "title": "Imaginary World (Chromes song)",
    "type": "MUSEWIKI_CHANGE",
    "uniqueId": "35999-219418",
    "user": "HyperChondriacMuser",
  },
  {
    "change_type": "edit",
    "comment": null,
    "created_at": 2025-06-07T10:01:23.000Z,
    "id": "219417",
    "newlen": 1117,
    "old_revid": 219416,
    "oldlen": 1103,
    "pageid": 35995,
    "title": "Chromes",
    "type": "MUSEWIKI_CHANGE",
    "uniqueId": "35995-219417",
    "user": "Casin0R0yale7",
  },
  {
    "change_type": "edit",
    "comment": null,
    "created_at": 2025-06-06T03:02:44.000Z,
    "id": "219416",
    "newlen": 1103,
    "old_revid": 219020,
    "oldlen": 1058,
    "pageid": 35995,
    "title": "Chromes",
    "type": "MUSEWIKI_CHANGE",
    "uniqueId": "35995-219416",
    "user": "NFSF1McLaren",
  },
  {
    "change_type": "edit",
    "comment": null,
    "created_at": 2025-06-01T20:56:27.000Z,
    "id": "219415",
    "newlen": 1720,
    "old_revid": 219407,
    "oldlen": 1717,
    "pageid": 36072,
    "title": "Helsinki Kulttuuritalo 2025 (gig)",
    "type": "MUSEWIKI_CHANGE",
    "uniqueId": "36072-219415",
    "user": "Luigiman09",
  },
  {
    "change_type": "edit",
    "comment": "Undo revision 219413 by [[Special:Contributions/Boombleeb|Boombleeb]] ([[User talk:Boombleeb|talk]])",
    "created_at": 2025-05-31T09:37:50.000Z,
    "id": "219414",
    "newlen": 4741,
    "old_revid": 219413,
    "oldlen": 4610,
    "pageid": 2029,
    "title": "Sydney Showground 2004 - 23rd (gig)",
    "type": "MUSEWIKI_CHANGE",
    "uniqueId": "2029-219414",
    "user": "MuseFanSam",
  },
  {
    "change_type": "edit",
    "comment": null,
    "created_at": 2025-05-30T14:24:34.000Z,
    "id": "219413",
    "newlen": 4610,
    "old_revid": 218775,
    "oldlen": 4741,
    "pageid": 2029,
    "title": "Sydney Showground 2004 - 23rd (gig)",
    "type": "MUSEWIKI_CHANGE",
    "uniqueId": "2029-219413",
    "user": "Boombleeb",
  },
  {
    "change_type": "edit",
    "comment": null,
    "created_at": 2025-05-30T07:21:18.000Z,
    "id": "219412",
    "newlen": 4028,
    "old_revid": 219405,
    "oldlen": 3387,
    "pageid": 36076,
    "title": "Summer Festivals Tour 2025",
    "type": "MUSEWIKI_CHANGE",
    "uniqueId": "36076-219412",
    "user": "Luigiman09",
  },
  {
    "change_type": "new",
    "comment": "Created page with "[[File:NOSAlive2025Poster.jpg|thumb|right|Festival lineup]] {{Gigbox | Venue = Passeio Maritimo de Algés | VenueRef = <ref>https://nosalive.com/en/</ref> | VenueTrans = Algés Riverside | Event = NOS Alive | EventRef = <ref>https://nosalive.com/en/</ref> | Date = 12{{supo|th}} July 2025 | DateRef = <ref>https://twitter.com/muse/status/1928054578706067920</ref> | Location = Lisbon | LocationRef = <ref>https://nosalive.com/en/how-to-get/</ref> | Country = Portugal | Songs..."",
    "created_at": 2025-05-30T06:57:52.000Z,
    "id": "219411",
    "newlen": 1666,
    "old_revid": 0,
    "oldlen": 0,
    "pageid": 36080,
    "title": "Lisbon Passeio Marítimo de Alges 2025 (gig)",
    "type": "MUSEWIKI_CHANGE",
    "uniqueId": "36080-219411",
    "user": "Luigiman09",
  },
  {
    "change_type": "log",
    "comment": "https://nosalive.com/en/muse-e-novidades-no-cartaz-do-nos-alive25/",
    "created_at": 2025-05-30T06:57:08.000Z,
    "id": "219410",
    "newlen": 0,
    "old_revid": 0,
    "oldlen": 0,
    "pageid": 36079,
    "title": "File:NOSAlive2025Poster.jpg",
    "type": "MUSEWIKI_CHANGE",
    "uniqueId": "36079-219410",
    "user": "Luigiman09",
  },
  {
    "change_type": "new",
    "comment": "Created page with "[[File:MadCool2025FestivalLineup.jpg|thumb|right|Festival lineup]] {{Gigbox | Venue = Iberdrola Music | VenueRef = <ref>https://madcoolfestival.es/</ref> | Event = Mad Cool Festival | EventRef = <ref>https://madcoolfestival.es/</ref> | Date = 10{{supo|th}} July 2025 | DateRef = <ref>https://www.instagram.com/p/DKPAzVWKMms/</ref> | Location = Madrid | LocationRef = <ref>https://www.instagram.com/p/DKPAzVWKMms/</ref> | Country = Spain | Songs = TBA | SongsRef =  | Support..."",
    "created_at": 2025-05-30T06:21:40.000Z,
    "id": "219409",
    "newlen": 1677,
    "old_revid": 0,
    "oldlen": 0,
    "pageid": 36078,
    "title": "Madrid Iberdrola Music 2025 (gig)",
    "type": "MUSEWIKI_CHANGE",
    "uniqueId": "36078-219409",
    "user": "Luigiman09",
  },
  {
    "change_type": "log",
    "comment": "https://twitter.com/madcoolfestival/status/1928039700960309670/photo/1",
    "created_at": 2025-05-30T06:18:29.000Z,
    "id": "219408",
    "newlen": 0,
    "old_revid": 0,
    "oldlen": 0,
    "pageid": 36077,
    "title": "File:MadCool2025FestivalLineup.jpg",
    "type": "MUSEWIKI_CHANGE",
    "uniqueId": "36077-219408",
    "user": "Luigiman09",
  },
  {
    "change_type": "edit",
    "comment": null,
    "created_at": 2025-05-29T17:30:17.000Z,
    "id": "219407",
    "newlen": 1717,
    "old_revid": 219391,
    "oldlen": 1728,
    "pageid": 36072,
    "title": "Helsinki Kulttuuritalo 2025 (gig)",
    "type": "MUSEWIKI_CHANGE",
    "uniqueId": "36072-219407",
    "user": "Luigiman09",
  },
  {
    "change_type": "edit",
    "comment": null,
    "created_at": 2025-05-29T17:29:48.000Z,
    "id": "219406",
    "newlen": 1482,
    "old_revid": 219401,
    "oldlen": 1443,
    "pageid": 36075,
    "title": "Gdynia Babie Doły 2025 (gig)",
    "type": "MUSEWIKI_CHANGE",
    "uniqueId": "36075-219406",
    "user": "Luigiman09",
  },
  {
    "change_type": "edit",
    "comment": "/* Gigs */",
    "created_at": 2025-05-29T17:27:31.000Z,
    "id": "219405",
    "newlen": 3387,
    "old_revid": 219393,
    "oldlen": 3146,
    "pageid": 36076,
    "title": "Summer Festivals Tour 2025",
    "type": "MUSEWIKI_CHANGE",
    "uniqueId": "36076-219405",
    "user": "Luigiman09",
  },
  {
    "change_type": "edit",
    "comment": null,
    "created_at": 2025-05-29T17:26:54.000Z,
    "id": "219404",
    "newlen": 1164,
    "old_revid": 219380,
    "oldlen": 923,
    "pageid": 36019,
    "title": "Template:Gigs/2025",
    "type": "MUSEWIKI_CHANGE",
    "uniqueId": "36019-219404",
    "user": "Luigiman09",
  },
  {
    "change_type": "edit",
    "comment": null,
    "created_at": 2025-05-28T08:35:24.000Z,
    "id": "219403",
    "newlen": 4162,
    "old_revid": 219402,
    "oldlen": 4157,
    "pageid": 4629,
    "title": "Tokyo International Forum 2007 (gig)",
    "type": "MUSEWIKI_CHANGE",
    "uniqueId": "4629-219403",
    "user": "Boombleeb",
  },
  {
    "change_type": "edit",
    "comment": null,
    "created_at": 2025-05-28T08:34:51.000Z,
    "id": "219402",
    "newlen": 4157,
    "old_revid": 217908,
    "oldlen": 4142,
    "pageid": 4629,
    "title": "Tokyo International Forum 2007 (gig)",
    "type": "MUSEWIKI_CHANGE",
    "uniqueId": "4629-219402",
    "user": "Boombleeb",
  },
  {
    "change_type": "edit",
    "comment": null,
    "created_at": 2025-05-22T06:59:24.000Z,
    "id": "219401",
    "newlen": 1443,
    "old_revid": 219387,
    "oldlen": 1375,
    "pageid": 36075,
    "title": "Gdynia Babie Doły 2025 (gig)",
    "type": "MUSEWIKI_CHANGE",
    "uniqueId": "36075-219401",
    "user": "Luigiman09",
  },
  {
    "change_type": "edit",
    "comment": null,
    "created_at": 2025-05-22T06:58:17.000Z,
    "id": "219400",
    "newlen": 1672,
    "old_revid": 219384,
    "oldlen": 1627,
    "pageid": 36065,
    "title": "Stockholm Gärdet 2025 (gig)",
    "type": "MUSEWIKI_CHANGE",
    "uniqueId": "36065-219400",
    "user": "Luigiman09",
  },
  {
    "change_type": "edit",
    "comment": null,
    "created_at": 2025-05-22T06:57:46.000Z,
    "id": "219399",
    "newlen": 2029,
    "old_revid": 219307,
    "oldlen": 1984,
    "pageid": 36018,
    "title": "Oslo Ekebergsletta 2025 (gig)",
    "type": "MUSEWIKI_CHANGE",
    "uniqueId": "36018-219399",
    "user": "Luigiman09",
  },
  {
    "change_type": "edit",
    "comment": null,
    "created_at": 2025-05-22T06:57:31.000Z,
    "id": "219398",
    "newlen": 1603,
    "old_revid": 219160,
    "oldlen": 1558,
    "pageid": 36024,
    "title": "Landgraaf Megaland 2025 (gig)",
    "type": "MUSEWIKI_CHANGE",
    "uniqueId": "36024-219398",
    "user": "Luigiman09",
  },
  {
    "change_type": "edit",
    "comment": null,
    "created_at": 2025-05-22T06:57:16.000Z,
    "id": "219397",
    "newlen": 1780,
    "old_revid": 219165,
    "oldlen": 1735,
    "pageid": 36032,
    "title": "Clisson Val de Moine 2025 (gig)",
    "type": "MUSEWIKI_CHANGE",
    "uniqueId": "36032-219397",
    "user": "Luigiman09",
  },
  {
    "change_type": "edit",
    "comment": null,
    "created_at": 2025-05-22T06:56:49.000Z,
    "id": "219396",
    "newlen": 1773,
    "old_revid": 219379,
    "oldlen": 1687,
    "pageid": 36027,
    "title": "Turku Artukaisten Airport 2025 (gig)",
    "type": "MUSEWIKI_CHANGE",
    "uniqueId": "36027-219396",
    "user": "Luigiman09",
  },
  {
    "change_type": "edit",
    "comment": null,
    "created_at": 2025-05-22T06:54:31.000Z,
    "id": "219395",
    "newlen": 1569,
    "old_revid": 212795,
    "oldlen": 1537,
    "pageid": 19211,
    "title": "Tour",
    "type": "MUSEWIKI_CHANGE",
    "uniqueId": "19211-219395",
    "user": "Luigiman09",
  },
  {
    "change_type": "edit",
    "comment": null,
    "created_at": 2025-05-22T06:53:34.000Z,
    "id": "219394",
    "newlen": 35605,
    "old_revid": 219248,
    "oldlen": 35575,
    "pageid": 33889,
    "title": "Will of the People World Tour",
    "type": "MUSEWIKI_CHANGE",
    "uniqueId": "33889-219394",
    "user": "Luigiman09",
  },
  {
    "change_type": "new",
    "comment": "Created page with "{{GigNav | Previous = [[Will of the People World Tour]] | Current = [[Summer Festivals Tour 2025]] | Next =  }}  This tour is a short tour of festivals around Europe, almost 3 years after the release of [[Will of the People (album)|Will of the People]]. It mostly stops in Nordic countries that Muse hasn't visited since the [[Simulation Theory (album)|Simulation Theory]] era, or the [[Drones (album)|Drones]] era, in the case of Sweden.  == Information ==  The Turku Artu..."",
    "created_at": 2025-05-22T06:53:13.000Z,
    "id": "219393",
    "newlen": 3146,
    "old_revid": 0,
    "oldlen": 0,
    "pageid": 36076,
    "title": "Summer Festivals Tour 2025",
    "type": "MUSEWIKI_CHANGE",
    "uniqueId": "36076-219393",
    "user": "Luigiman09",
  },
  {
    "change_type": "edit",
    "comment": null,
    "created_at": 2025-05-18T21:23:09.000Z,
    "id": "219392",
    "newlen": 1457,
    "old_revid": 215967,
    "oldlen": 1426,
    "pageid": 9030,
    "title": "MuseTV",
    "type": "MUSEWIKI_CHANGE",
    "uniqueId": "9030-219392",
    "user": "Nepomuse",
  },
  {
    "change_type": "edit",
    "comment": "price",
    "created_at": 2025-05-18T10:47:37.000Z,
    "id": "219391",
    "newlen": 1728,
    "old_revid": 219390,
    "oldlen": 1719,
    "pageid": 36072,
    "title": "Helsinki Kulttuuritalo 2025 (gig)",
    "type": "MUSEWIKI_CHANGE",
    "uniqueId": "36072-219391",
    "user": "Ncla",
  },
  {
    "change_type": "edit",
    "comment": "Doors",
    "created_at": 2025-05-18T10:36:11.000Z,
    "id": "219390",
    "newlen": 1719,
    "old_revid": 219382,
    "oldlen": 1709,
    "pageid": 36072,
    "title": "Helsinki Kulttuuritalo 2025 (gig)",
    "type": "MUSEWIKI_CHANGE",
    "uniqueId": "36072-219390",
    "user": "Ncla",
  },
  {
    "change_type": "edit",
    "comment": null,
    "created_at": 2025-05-17T12:06:16.000Z,
    "id": "219389",
    "newlen": 1935,
    "old_revid": 219388,
    "oldlen": 1927,
    "pageid": 23339,
    "title": "User:Ncla",
    "type": "MUSEWIKI_CHANGE",
    "uniqueId": "23339-219389",
    "user": "Ncla",
  },
  {
    "change_type": "edit",
    "comment": "😎",
    "created_at": 2025-05-17T12:03:41.000Z,
    "id": "219388",
    "newlen": 1927,
    "old_revid": 219378,
    "oldlen": 1838,
    "pageid": 23339,
    "title": "User:Ncla",
    "type": "MUSEWIKI_CHANGE",
    "uniqueId": "23339-219388",
    "user": "Ncla",
  },
  {
    "change_type": "new",
    "comment": "Created page with "[[File:Opener2025Announcement.jpg|thumb|right|Muse announcement]] {{Gigbox | Venue = Babie Doły Military Airport | VenueRef = <ref>https://www.muse.mu/tour-date/opener-festival-gdynia-poland</ref> | Event = Open'er Festival | EventRef = <ref>https://www.instagram.com/p/DJGuqP7tpE8/</ref> | Date = 4{{supo|th}} July 2025 | DateRef = <ref>https://www.instagram.com/p/DJEJ3jUoRMu/</ref> | Location = Gdynia | LocationRef = <ref>https://www.muse.mu/tour-date/opener-festival-gd..."",
    "created_at": 2025-05-17T07:09:16.000Z,
    "id": "219387",
    "newlen": 1375,
    "old_revid": 0,
    "oldlen": 0,
    "pageid": 36075,
    "title": "Gdynia Babie Doły 2025 (gig)",
    "type": "MUSEWIKI_CHANGE",
    "uniqueId": "36075-219387",
    "user": "Luigiman09",
  },
  {
    "change_type": "log",
    "comment": "https://www.instagram.com/p/DJEJ3jUoRMu/",
    "created_at": 2025-05-17T07:09:02.000Z,
    "id": "219386",
    "newlen": 0,
    "old_revid": 0,
    "oldlen": 0,
    "pageid": 36074,
    "title": "File:Opener2025Announcement.jpg",
    "type": "MUSEWIKI_CHANGE",
    "uniqueId": "36074-219386",
    "user": "Luigiman09",
  },
  {
    "change_type": "edit",
    "comment": null,
    "created_at": 2025-05-17T06:38:22.000Z,
    "id": "219385",
    "newlen": 2458,
    "old_revid": 219371,
    "oldlen": 2456,
    "pageid": 36069,
    "title": "Istanbul Festival Park Yenikapi 2025 (gig)",
    "type": "MUSEWIKI_CHANGE",
    "uniqueId": "36069-219385",
    "user": "Luigiman09",
  },
  {
    "change_type": "edit",
    "comment": null,
    "created_at": 2025-05-17T06:37:23.000Z,
    "id": "219384",
    "newlen": 1627,
    "old_revid": 219306,
    "oldlen": 1435,
    "pageid": 36065,
    "title": "Stockholm Gärdet 2025 (gig)",
    "type": "MUSEWIKI_CHANGE",
    "uniqueId": "36065-219384",
    "user": "Luigiman09",
  },
  {
    "change_type": "log",
    "comment": "https://www.instagram.com/p/DGknZ7lC8sh/",
    "created_at": 2025-05-17T06:33:21.000Z,
    "id": "219383",
    "newlen": 0,
    "old_revid": 0,
    "oldlen": 0,
    "pageid": 36073,
    "title": "File:STHLM25 MuseAnnouncement-2.jpg",
    "type": "MUSEWIKI_CHANGE",
    "uniqueId": "36073-219383",
    "user": "Luigiman09",
  },
  {
    "change_type": "new",
    "comment": "Created page with "[[File:Helsinki2025gigannouncement.jpg|thumb|Announcement picture posted on social media]] {{Gigbox | Venue = Kulttuuritalo | VenueRef = <ref>https://en.wikipedia.org/wiki/Kulttuuritalo</ref> | VenueTrans = House of Culture | Date = 12{{supo|th}} June 2025 | DateRef = <ref>https://www.muse.mu/tour-date/house-culture-helsinki-finland</ref> | Location = Helsinki, ES | LocationRef = <ref>https://www.muse.mu/tour-date/house-culture-helsinki-finland</ref> | Country = Finland..."",
    "created_at": 2025-05-17T06:21:49.000Z,
    "id": "219382",
    "newlen": 1709,
    "old_revid": 0,
    "oldlen": 0,
    "pageid": 36072,
    "title": "Helsinki Kulttuuritalo 2025 (gig)",
    "type": "MUSEWIKI_CHANGE",
    "uniqueId": "36072-219382",
    "user": "Luigiman09",
  },
  {
    "change_type": "log",
    "comment": "https://www.instagram.com/p/DJlhabeNmGM/",
    "created_at": 2025-05-17T06:16:27.000Z,
    "id": "219381",
    "newlen": 0,
    "old_revid": 0,
    "oldlen": 0,
    "pageid": 36071,
    "title": "File:Helsinki2025gigannouncement.jpg",
    "type": "MUSEWIKI_CHANGE",
    "uniqueId": "36071-219381",
    "user": "Luigiman09",
  },
  {
    "change_type": "edit",
    "comment": null,
    "created_at": 2025-05-17T05:37:07.000Z,
    "id": "219380",
    "newlen": 923,
    "old_revid": 219364,
    "oldlen": 734,
    "pageid": 36019,
    "title": "Template:Gigs/2025",
    "type": "MUSEWIKI_CHANGE",
    "uniqueId": "36019-219380",
    "user": "Luigiman09",
  },
  {
    "change_type": "edit",
    "comment": null,
    "created_at": 2025-05-17T05:33:07.000Z,
    "id": "219379",
    "newlen": 1687,
    "old_revid": 219360,
    "oldlen": 1716,
    "pageid": 36027,
    "title": "Turku Artukaisten Airport 2025 (gig)",
    "type": "MUSEWIKI_CHANGE",
    "uniqueId": "36027-219379",
    "user": "Luigiman09",
  },
  {
    "change_type": "edit",
    "comment": "update gig list with STHLM fields",
    "created_at": 2025-04-27T08:44:41.000Z,
    "id": "219378",
    "newlen": 1838,
    "old_revid": 216640,
    "oldlen": 1758,
    "pageid": 23339,
    "title": "User:Ncla",
    "type": "MUSEWIKI_CHANGE",
    "uniqueId": "23339-219378",
    "user": "Ncla",
  },
  {
    "change_type": "edit",
    "comment": "fixed last shadow puppets name",
    "created_at": 2025-04-20T21:49:03.000Z,
    "id": "219377",
    "newlen": 2930,
    "old_revid": 218905,
    "oldlen": 2922,
    "pageid": 31901,
    "title": "The Jaded Hearts Club",
    "type": "MUSEWIKI_CHANGE",
    "uniqueId": "31901-219377",
    "user": "Tdrunner95",
  },
  {
    "change_type": "edit",
    "comment": "/* Rack (right) */",
    "created_at": 2025-04-03T21:33:38.000Z,
    "id": "219376",
    "newlen": 14080,
    "old_revid": 219375,
    "oldlen": 14057,
    "pageid": 1430,
    "title": "Guitar effects & Amplifiers",
    "type": "MUSEWIKI_CHANGE",
    "uniqueId": "1430-219376",
    "user": "Onlineguitarcovers",
  },
  {
    "change_type": "edit",
    "comment": "/* Rack (right) */",
    "created_at": 2025-04-03T21:32:22.000Z,
    "id": "219375",
    "newlen": 14057,
    "old_revid": 219374,
    "oldlen": 14032,
    "pageid": 1430,
    "title": "Guitar effects & Amplifiers",
    "type": "MUSEWIKI_CHANGE",
    "uniqueId": "1430-219375",
    "user": "Onlineguitarcovers",
  },
  {
    "change_type": "edit",
    "comment": "/* Rack (right) */",
    "created_at": 2025-04-03T21:31:10.000Z,
    "id": "219374",
    "newlen": 14032,
    "old_revid": 219373,
    "oldlen": 13992,
    "pageid": 1430,
    "title": "Guitar effects & Amplifiers",
    "type": "MUSEWIKI_CHANGE",
    "uniqueId": "1430-219374",
    "user": "Onlineguitarcovers",
  },
  {
    "change_type": "edit",
    "comment": "/*  Will Of The People Tour */",
    "created_at": 2025-04-03T21:28:54.000Z,
    "id": "219373",
    "newlen": 13992,
    "old_revid": 219033,
    "oldlen": 13935,
    "pageid": 1430,
    "title": "Guitar effects & Amplifiers",
    "type": "MUSEWIKI_CHANGE",
    "uniqueId": "1430-219373",
    "user": "Onlineguitarcovers",
  },
  {
    "change_type": "log",
    "comment": null,
    "created_at": 2025-04-03T21:28:36.000Z,
    "id": "219372",
    "newlen": 0,
    "old_revid": 0,
    "oldlen": 0,
    "pageid": 36070,
    "title": "File:Screenshot 2025-04-03 222731.png",
    "type": "MUSEWIKI_CHANGE",
    "uniqueId": "36070-219372",
    "user": "Onlineguitarcovers",
  },
]
`;
